src/pytest_bdd/gherkin_parser.py

```diff
@@ -276,11 +316,30 @@ def get_gherkin_document(abs_filename: str = None, encoding: str = "utf-8") -> G
     try:
         gherkin_data = Parser().parse(TokenScanner(feature_file_text))
     except CompositeParserException as e:
-        raise exceptions.FeatureError(
-            e.args[0],
-            e.errors[0].location["line"],
-            linecache.getline(abs_filename, e.errors[0].location["line"]).rstrip("\n"),
-            abs_filename,
-        ) from e
-
+        message = e.args[0]
+        line = e.errors[0].location["line"]
+        line_content = linecache.getline(abs_filename, e.errors[0].location["line"]).rstrip("\n")
+        filename = abs_filename
+        gherkin_error_handler = GherkinParserErrorHandler()
+        gherkin_error_handler(message, line, line_content, filename)
+        # If no patterns matched, raise a generic GherkinParserError
+        raise exceptions.GherkinParseError(f"Unknown parsing error: {message}", line, line_content, filename)
+
+    # At this point, the `gherkin_data` should be valid if no exception was raised
     return GherkinDocument.from_dict(gherkin_data)
+
+
+class GherkinParserErrorHandler:
```

Comment:
this can just be a function right?

Reply:
Yep - I started making it more complicated than need be, but the current logic makes sense to be a function.  Updated.